### PR TITLE
Auto Case Update Refactor

### PR DESCRIPTION
**Pull request by gcapalbo**
Details here:
https://docs.google.com/document/d/1M3OWqYyTQKBGvB_xEmkB9bsTH1ZvCGNXqIs3nvRQxR0/edit#

Demo here:
https://staging.commcarehq.org/a/gc-staging/data/edit/automatic_updates/

buddy @mkangia 
cc @calellowitz for updates to rule running processes (this extends the work you had done)

### CODE REVIEW

File: corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_actions.js
Comment (dimagimon) [inline corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_actions.js]: 'CaseRuleActions' is defined but never used (no-unused-vars)

File: corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_actions.js
Comment (dimagimon) [inline corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_actions.js]: 'self' is defined but never used (no-unused-vars)

File: corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_criteria.js
Comment (dimagimon) [inline corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_criteria.js]: 'CaseRuleCriteria' is defined but never used (no-unused-vars)

File: corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_criteria.js
Comment (dimagimon) [inline corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_criteria.js]: 'obj' is already defined (no-redeclare)

**Comment by gcapalbo** [comment]
fine to review, but i want to add tests before merging

**Review by mkangia — APPROVED** [review]
I don't have any context on this and would take quite a lot of digging in. Will defer to @calellowitz  for review or if you want to add some more reviewers @gcapalbo

File: corehq/apps/data_interfaces/templates/data_interfaces/edit_case_rule.html
Comment (calellowitz) [inline corehq/apps/data_interfaces/templates/data_interfaces/edit_case_rule.html]: same comment as above

**Review by calellowitz — CHANGES_REQUESTED** [review]
left one real change request, hopefully it came through and isnt lost to an old diff

**Comment by calellowitz** [comment]
Looks good besides the one comment, though admittedly it was large enough of a PR that I was not able to read every part super closely. Let me know if github hides the comment since you ended up modifying the section i made it on.

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/apps/data_interfaces/models.py, line 956 - F821 undefined name 'domain'
* corehq/apps/data_interfaces/models.py, line 964 - F821 undefined name 'since'
* corehq/apps/data_interfaces/models.py, line 982 - F821 undefined name 'FormAccessors'
* corehq/apps/data_interfaces/models.py, line 985 - F821 undefined name 'instance'
* corehq/apps/data_interfaces/tests/test_case_rule_migration.py, line 1 - F401 'json' imported but unused

File: corehq/apps/data_interfaces/tests/test_case_rule_migration.py
Comment (dimagimon) [inline corehq/apps/data_interfaces/tests/test_case_rule_migration.py]: F401 'json' imported but unused

File: corehq/apps/data_interfaces/static/data_interfaces/js/case_rule.js
Comment (dimagimon) [inline corehq/apps/data_interfaces/static/data_interfaces/js/case_rule.js]: 'event' is defined but never used (no-unused-vars)

File: corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_actions.js
Comment (dimagimon) [inline corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_actions.js]: 'self' is defined but never used (no-unused-vars)

File: corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_actions.js
Comment (dimagimon) [inline corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_actions.js]: Missing semicolon. (semi)

File: corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_criteria.js
Comment (dimagimon) [inline corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_criteria.js]: Missing semicolon. (semi)

File: corehq/apps/data_interfaces/tests/test_auto_case_updates.py
Comment (dimagimon) [inline corehq/apps/data_interfaces/tests/test_auto_case_updates.py]: F811 redefinition of unused 'test_case_property_equal' from line 619

File: corehq/apps/data_interfaces/tests/test_auto_case_updates.py
Comment (dimagimon) [inline corehq/apps/data_interfaces/tests/test_auto_case_updates.py]: F841 local variable 'rule4' is assigned to but never used

File: corehq/apps/data_interfaces/tests/test_auto_case_updates.py
Comment (dimagimon) [inline corehq/apps/data_interfaces/tests/test_auto_case_updates.py]: F841 local variable 'rule5' is assigned to but never used

File: corehq/apps/data_interfaces/tests/test_auto_case_updates.py
Comment (dimagimon) [inline corehq/apps/data_interfaces/tests/test_auto_case_updates.py]: F841 local variable 'rule4' is assigned to but never used

**Comment by gcapalbo** [comment]
@calellowitz I externalized all the js and added some more tests - this should be good to go now.

(Leaving the don't merge label just so it doesn't get accidentally deployed overnight or something)

**Review by calellowitz — APPROVED** [review]

**Pull request merged**
